lib/recursive-methods.rb

```diff
@@ -1,47 +1,86 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) because n number of multplications
+# Space complexity: O(n) because n-1 in activation chain
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    # recursion here
+    if n == 0
+        return 1
+    elsif n == -1
+        raise ArgumentError, "unable to calculate because of negative number"
+    else
+        return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+#------------------------------------------------------------------------
+# Time complexity: O(n) because n is = to number being passed
+# Space complexity: O(n) because n is = to number being passed
+
 def reverse(s)
```

Comment:
The space complexity is O(n^2) because you're making a new array with each recursive call.